File: backend/calculators/foundation_bearing.py
"""GB 50007—2011 天然地基承载力与矩形基础底面压力验算。"""

import math
from dataclasses import dataclass
from typing import Optional


SOIL_CORRECTION_PRESETS = {
    "silt_muck_fill_soft_clay": ("淤泥、人工填土或软黏性土", 0.0, 1.0),
    "red_clay_wet": ("红黏土（含水比 αw＞0.8）", 0.0, 1.2),
    "red_clay_dry": ("红黏土（含水比 αw≤0.8）", 0.15, 1.4),
    "compacted_silt": ("大面积压实粉土", 0.0, 1.5),
    "compacted_gravel": ("大面积压实级配砂石", 0.0, 2.0),
    "silt_clay_ge10": ("粉土（黏粒含量≥10%）", 0.3, 1.5),
    "silt_clay_lt10": ("粉土（黏粒含量＜10%）", 0.5, 2.0),
    "cohesive_firm": ("e、IL 均＜0.85 的黏性土", 0.3, 1.6),
    "fine_sand": ("粉砂、细砂（不含很湿/饱和稍密）", 2.0, 3.0),
    "coarse_soil": ("中粗砂、砾砂及碎石土", 3.0, 4.4),
}


@dataclass
class FoundationBearingInput:
    b: float
    l: float
    d: float
    fak: float
    gamma: float
    gamma_m: float
    Fk: float
    Mx: float = 0.0
    My: float = 0.0
    Gk: Optional[float] = None
    foundation_weight_pressure: float = 20.0
    soil_category: str = "cohesive_firm"
    eta_b: Optional[float] = None
    eta_d: Optional[float] = None


def _positive(value: float, label: str) -> float:
    number = float(value)
    if not math.isfinite(number) or number <= 0:
        raise ValueError(f"请填写有效的{label}")
    return number


def _nonnegative(value: float, label: str) -> float:
    number = float(value)
    if not math.isfinite(number) or number < 0:
        raise ValueError(f"{label}不能小于 0")
    return number


def calculate_foundation_bearing(inp: FoundationBearingInput) -> dict:
    b = _positive(inp.b, "基础宽度 b")
    length = _positive(inp.l, "基础长度 l")
    d = _nonnegative(inp.d, "基础埋深 d")
    fak = _positive(inp.fak, "地基承载力特征值 fak")
    gamma = _positive(inp.gamma, "基底以下土重度 γ")
    gamma_m = _positive(inp.gamma_m, "基底以上土平均重度 γm")
    fk = _nonnegative(inp.Fk, "上部竖向力 Fk")
    mx, my = float(inp.Mx), float(inp.My)
    if not math.isfinite(mx) or not math.isfinite(my):
        raise ValueError("基础底面弯矩必须为有效数字")

    if inp.soil_category == "custom":
        if inp.eta_b is None or inp.eta_d is None:
            raise ValueError("自定义土类必须填写 ηb 和 ηd")
        soil_label = "自定义修正系数"
        eta_b = _nonnegative(inp.eta_b, "宽度修正系数 ηb")
        eta_d = _nonnegative(inp.eta_d, "深度修正系数 ηd")
    else:
        if inp.soil_category not in SOIL_CORRECTION_PRESETS:
            raise ValueError("不支持的持力层土类别")
        soil_label, eta_b, eta_d = SOIL_CORRECTION_PRESETS[inp.soil_category]

    area = b * length
    if inp.Gk is None:
        foundation_weight_pressure = _positive(inp.foundation_weight_pressure, "单位面积基础及覆土重")
        gk = foundation_weight_pressure * area
        weight_source = "auto"
    else:
        foundation_weight_pressure = None
        gk = _nonnegative(inp.Gk, "基础及覆土重 Gk")
        weight_source = "manual"
    vertical = fk + gk
    if vertical <= 0:
        raise ValueError("竖向合力 Fk＋Gk 必须大于 0")

    b_corr = min(6.0, max(3.0, min(b, length)))
    d_corr = max(0.5, d)
    width_increment = eta_b * gamma * (b_corr - 3.0)
    depth_increment = eta_d * gamma_m * (d_corr - 0.5)
    fa = fak + width_increment + depth_increment
    wx = b * length ** 2 / 6.0
    wy = length * b ** 2 / 6.0
    pk = vertical / area
    tx, ty = abs(mx) / wx, abs(my) / wy
    pmax_linear = pk + tx + ty
    pmin_linear = pk - tx - ty

    tolerance = 1e-10
    has_mx, has_my = abs(mx) > tolerance, abs(my) > tolerance
    biaxial = has_mx and has_my
    full_contact = pmin_linear >= -tolerance
    supported, stable = True, True
    pressure_mode = "linear_full_contact" if has_mx or has_my else "uniform"
    pmax, pmin = pmax_linear, max(pmin_linear, 0.0)
    contact_width = b if has_my else length
    eccentricity = 0.0

    if not full_contact:
        if biaxial:
            supported = False
            pressure_mode = "biaxial_separation_review"
            pmax, pmin, contact_width = None, pmin_linear, None
        else:
            moment = max(abs(mx), abs(my))
            direction = b if has_my else length
            perpendicular = length if has_my else b
            eccentricity = moment / vertical
            a = direction / 2.0 - eccentricity
            pressure_mode = "uniaxial_triangular_contact"
            if a <= 0:
                supported, stable = False, False
                pressure_mode = "resultant_outside_base"
                pmax, pmin, contact_width = None, None, 0.0
            else:
                pmax = 2.0 * vertical / (3.0 * perpendicular * a)
                pmin = 0.0
                contact_width = 3.0 * a

    mean_pass = pk <= fa + tolerance
    edge_pass = pmax <= 1.2 * fa + tolerance if supported and stable and pmax is not None else None
    overall_pass = mean_pass and edge_pass is True and stable and supported
    status = "pass" if overall_pass else "fail"
    if not supported and stable:
        status = "review"
    if not stable:
        status = "unstable"

    corners = {
        "top_left": pk + mx / wx - my / wy,
        "top_right": pk + mx / wx + my / wy,
        "bottom_left": pk - mx / wx - my / wy,
        "bottom_right": pk - mx / wx + my / wy,
    }
    rounded = lambda value: None if value is None else round(value + 1e-12, 3)
    return {
        "soil_category": inp.soil_category,
        "soil_label": soil_label,
        "eta_b": rounded(eta_b),
        "eta_d": rounded(eta_d),
        "b_correction": rounded(b_corr),
        "d_correction": rounded(d_corr),
        "area": rounded(area),
        "Wx": rounded(wx),
        "Wy": rounded(wy),
        "Gk": rounded(gk),
        "N": rounded(vertical),
        "fa": rounded(fa),
        "width_increment": rounded(width_increment),
        "depth_increment": rounded(depth_increment),
        "pk": rounded(pk),
        "pmax": rounded(pmax),
        "pmin": rounded(pmin),
        "pmax_linear": rounded(pmax_linear),
        "pmin_linear": rounded(pmin_linear),
        "eccentricity": rounded(eccentricity),
        "contact_width": rounded(contact_width),
        "pressure_mode": pressure_mode,
        "full_contact": full_contact,
        "supported": supported,
        "stable": stable,
        "mean_pass": mean_pass,
        "edge_pass": edge_pass,
        "overall_pass": overall_pass,
        "mean_utilization": rounded(pk / fa),
        "edge_utilization": rounded(pmax / (1.2 * fa)) if pmax is not None else None,
        "corners": {key: rounded(value) for key, value in corners.items()},
        "weight_source": weight_source,
        "foundation_weight_pressure": foundation_weight_pressure,
        "status": status,
    }

File: utils/calculators/foundation-bearing.js
import { roundTo, requirePositive } from './common.js'

export const SOIL_CORRECTION_PRESETS = {
  silt_muck_fill_soft_clay: { label: '淤泥、人工填土或软黏性土', eta_b: 0, eta_d: 1.0 },
  red_clay_wet: { label: '红黏土（含水比 αw＞0.8）', eta_b: 0, eta_d: 1.2 },
  red_clay_dry: { label: '红黏土（含水比 αw≤0.8）', eta_b: 0.15, eta_d: 1.4 },
  compacted_silt: { label: '大面积压实粉土', eta_b: 0, eta_d: 1.5 },
  compacted_gravel: { label: '大面积压实级配砂石', eta_b: 0, eta_d: 2.0 },
  silt_clay_ge10: { label: '粉土（黏粒含量≥10%）', eta_b: 0.3, eta_d: 1.5 },
  silt_clay_lt10: { label: '粉土（黏粒含量＜10%）', eta_b: 0.5, eta_d: 2.0 },
  cohesive_firm: { label: 'e、IL 均＜0.85 的黏性土', eta_b: 0.3, eta_d: 1.6 },
  fine_sand: { label: '粉砂、细砂（不含很湿/饱和稍密）', eta_b: 2.0, eta_d: 3.0 },
  coarse_soil: { label: '中粗砂、砾砂及碎石土', eta_b: 3.0, eta_d: 4.4 },
  custom: { label: '自定义修正系数', eta_b: null, eta_d: null }
}

const finiteNonnegative = (value, label) => {
  const number = Number(value ?? 0)
  if (!Number.isFinite(number) || number < 0) throw new Error(`${label}不能小于 0`)
  return number
}

function normalize(raw = {}) {
  const b = requirePositive(raw.b, '基础宽度 b')
  const l = requirePositive(raw.l, '基础长度 l')
  const d = finiteNonnegative(raw.d, '基础埋深 d')
  const fak = requirePositive(raw.fak, '地基承载力特征值 fak')
  const gamma = requirePositive(raw.gamma, '基底以下土重度 γ')
  const gammaM = requirePositive(raw.gamma_m, '基底以上土平均重度 γm')
  const Fk = finiteNonnegative(raw.Fk, '上部竖向力 Fk')
  const Mx = Number(raw.Mx ?? 0)
  const My = Number(raw.My ?? 0)
  if (!Number.isFinite(Mx) || !Number.isFinite(My)) throw new Error('基础底面弯矩必须为有效数字')

  const soilCategory = raw.soil_category || 'cohesive_firm'
  const preset = SOIL_CORRECTION_PRESETS[soilCategory]
  if (!preset) throw new Error('不支持的持力层土类别')
  if (soilCategory === 'custom' && (raw.eta_b === undefined || raw.eta_b === null || raw.eta_b === '' ||
    raw.eta_d === undefined || raw.eta_d === null || raw.eta_d === '')) {
    throw new Error('自定义土类必须填写 ηb 和 ηd')
  }
  const etaB = finiteNonnegative(soilCategory === 'custom' ? raw.eta_b : preset.eta_b, '宽度修正系数 ηb')
  const etaD = finiteNonnegative(soilCategory === 'custom' ? raw.eta_d : preset.eta_d, '深度修正系数 ηd')

  const area = b * l
  let weightSource = 'manual'
  let foundationWeightPressure = null
  let Gk
  if (raw.Gk !== undefined && raw.Gk !== null && raw.Gk !== '') {
    Gk = finiteNonnegative(raw.Gk, '基础及覆土重 Gk')
  } else {
    foundationWeightPressure = requirePositive(raw.foundation_weight_pressure ?? 20, '单位面积基础及覆土重')
    Gk = foundationWeightPressure * area
    weightSource = 'auto'
  }
  const N = Fk + Gk
  if (N <= 0) throw new Error('竖向合力 Fk＋Gk 必须大于 0')

  return {
    b, l, d, fak, gamma, gamma_m: gammaM, Fk, Gk, Mx, My, N, area,
    soil_category: soilCategory, soil_label: preset.label, eta_b: etaB, eta_d: etaD,
    weight_source: weightSource, foundation_weight_pressure: foundationWeightPressure
  }
}

export function calculateFoundationBearing(raw = {}) {
  const input = normalize(raw)
  const bCorrection = Math.min(6, Math.max(3, Math.min(input.b, input.l)))
  const dCorrection = Math.max(0.5, input.d)
  const widthIncrement = input.eta_b * input.gamma * (bCorrection - 3)
  const depthIncrement = input.eta_d * input.gamma_m * (dCorrection - 0.5)
  const fa = input.fak + widthIncrement + depthIncrement

  const Wx = input.b * input.l ** 2 / 6
  const Wy = input.l * input.b ** 2 / 6
  const tx = Math.abs(input.Mx) / Wx
  const ty = Math.abs(input.My) / Wy
  const pk = input.N / input.area
  const pmaxLinear = pk + tx + ty
  const pminLinear = pk - tx - ty
  const tolerance = 1e-10
  const hasMx = Math.abs(input.Mx) > tolerance
  const hasMy = Math.abs(input.My) > tolerance
  const biaxial = hasMx && hasMy
  let fullContact = pminLinear >= -tolerance
  let supported = true
  let stable = true
  let pressureMode = hasMx || hasMy ? 'linear_full_contact' : 'uniform'
  let pmax = pmaxLinear
  let pmin = Math.max(pminLinear, 0)
  let contactWidth = hasMy ? input.b : input.l
  let eccentricity = 0

  if (!fullContact) {
    if (biaxial) {
      supported = false
      pressureMode = 'biaxial_separation_review'
      pmax = null
      pmin = pminLinear
      contactWidth = null
    } else {
      const moment = Math.max(Math.abs(input.Mx), Math.abs(input.My))
      const directionSize = hasMy ? input.b : input.l
      const perpendicularSize = hasMy ? input.l : input.b
      eccentricity = moment / input.N
      const a = directionSize / 2 - eccentricity
      pressureMode = 'uniaxial_triangular_contact'
      if (a <= 0) {
        supported = false
        stable = false
        pressureMode = 'resultant_outside_base'
        pmax = null
        pmin = null
        contactWidth = 0
      } else {
        pmax = 2 * input.N / (3 * perpendicularSize * a)
        pmin = 0
        contactWidth = 3 * a
      }
    }
  }

  const meanPass = pk <= fa + tolerance
  const edgePass = supported && stable && pmax !== null ? pmax <= 1.2 * fa + tolerance : null
  const overallPass = meanPass && edgePass === true && stable && supported
  let status = overallPass ? 'pass' : 'fail'
  if (!supported && stable) status = 'review'
  if (!stable) status = 'unstable'

  const corners = {
    top_left: pk + input.Mx / Wx - input.My / Wy,
    top_right: pk + input.Mx / Wx + input.My / Wy,
    bottom_left: pk - input.Mx / Wx - input.My / Wy,
    bottom_right: pk - input.Mx / Wx + input.My / Wy
  }
  const steps = [
    `基础底面积 A = b·l = ${input.b.toFixed(3)}×${input.l.toFixed(3)} = ${input.area.toFixed(3)} m²`,
    `修正用宽度 b₀ = ${bCorrection.toFixed(3)} m，修正用埋深 d₀ = ${dCorrection.toFixed(3)} m`,
    `fₐ = fₐₖ＋ηbγ(b₀－3)＋ηdγm(d₀－0.5) = ${fa.toFixed(3)} kPa`,
    `竖向合力 Nₖ = Fₖ＋Gₖ = ${input.N.toFixed(3)} kN`,
    `平均压力 pₖ = Nₖ/A = ${pk.toFixed(3)} kPa`,
    `基础底面抵抗矩 Wₓ = ${Wx.toFixed(3)} m³，Wᵧ = ${Wy.toFixed(3)} m³`,
    pressureMode === 'uniaxial_triangular_contact'
      ? `单向大偏心：e = ${eccentricity.toFixed(3)} m，受压宽度 3a = ${contactWidth.toFixed(3)} m，pₖ,max = ${pmax.toFixed(3)} kPa`
      : pressureMode === 'biaxial_separation_review'
        ? '双向偏心出现基底脱空，规范简化线性压力公式不再适用，需进行专门的接触压力分析'
        : pressureMode === 'resultant_outside_base'
          ? '竖向合力作用线位于基础底面以外，基础存在倾覆失稳风险'
          : `基底保持全截面受压，pₖ,max = ${pmax.toFixed(3)} kPa，pₖ,min = ${pmin.toFixed(3)} kPa`,
    `平均压力验算：pₖ ${meanPass ? '≤' : '＞'} fₐ`,
    edgePass === null
      ? '边缘最大压力验算：当前工况不适用本页简化公式'
      : `边缘压力验算：pₖ,max ${edgePass ? '≤' : '＞'} 1.2fₐ`
  ]

  const roundedCorners = Object.fromEntries(Object.entries(corners).map(([key, value]) => [key, roundTo(value)]))
  return {
    data: {
      soil_category: input.soil_category,
      soil_label: input.soil_label,
      eta_b: roundTo(input.eta_b),
      eta_d: roundTo(input.eta_d),
      b_correction: roundTo(bCorrection),
      d_correction: roundTo(dCorrection),
      area: roundTo(input.area),
      Wx: roundTo(Wx),
      Wy: roundTo(Wy),
      Gk: roundTo(input.Gk),
      N: roundTo(input.N),
      fa: roundTo(fa),
      width_increment: roundTo(widthIncrement),
      depth_increment: roundTo(depthIncrement),
      pk: roundTo(pk),
      pmax: pmax === null ? null : roundTo(pmax),
      pmin: pmin === null ? null : roundTo(pmin),
      pmax_linear: roundTo(pmaxLinear),
      pmin_linear: roundTo(pminLinear),
      eccentricity: roundTo(eccentricity),
      contact_width: contactWidth === null ? null : roundTo(contactWidth),
      pressure_mode: pressureMode,
      full_contact: fullContact,
      supported,
      stable,
      mean_pass: meanPass,
      edge_pass: edgePass,
      overall_pass: overallPass,
      mean_utilization: roundTo(pk / fa),
      edge_utilization: pmax === null ? null : roundTo(pmax / (1.2 * fa)),
      corners: roundedCorners,
      steps,
      assumptions: [
        '采用作用标准组合，基础底面为矩形，输入弯矩为基础底面形心处弯矩',
        'fₐₖ 应由勘察、载荷试验、原位测试或工程经验综合确定，本工具不替代勘察',
        '未包含软弱下卧层、沉降、抗滑、抗倾覆、冲切和基础配筋验算'
      ],
      weight_source: input.weight_source,
      foundation_weight_pressure: input.foundation_weight_pressure
    },
    message: overallPass ? '地基承载力验算满足' : status === 'review' ? '需要进行接触压力专项复核' : '地基承载力验算不满足',
    status
  }
}
